### ISSUE TITLE
How to compute any actual quantity of the coarse-grained kinetics with employing the fuzzy memberships?

### ISSUE THREAD

**Issue by ShenWenHuibit** [deeptime-ml/deeptime#279]
`bayesian_hmm.gather_stats('transition_model/mfpt', A=[hmm_4.samples[0].metastable_sets[i]], B=hmm_4.samples[0].metastable_sets[j]).mean`

After executing the above code it gives me error:
`ValueError: Chosen set contains states that are not included in the active set.`

I don't know how to choose active set, please help me, thanks.

**Comment by clonker** [comment]
It most likely means that due to the sampling you have produced a few hmm's who live on different states (or at least not a subset) of the prior's states. I suggest you subselect the samples accordingly. It might even be enough to call `bhmm_largest = bhmm.submodel_largest()`

**Comment by ShenWenHuibit** [comment]
> It most likely means that due to the sampling you have produced a few hmm's who live on different states (or at least not a subset) of the prior's states. I suggest you subselect the samples accordingly. It might even be enough to call `bhmm_largest = bhmm.submodel_largest()`

When I run the command `bhmm_largest = hmm.submodel_largest()`, it gives me another error:

`---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[278], line 1
----> 1 bhmm_largest = hmm_4.submodel_largest()

File [~/whshen/anaconda3/envs/workshop/lib/python3.9/site-packages/deeptime/markov/hmm/_bayesian_hmm.py:65](https://vscode-remote+ssh-002dremote-002b10-002e2-002e1-002e240.vscode-resource.vscode-cdn.net/home/xhshi/whshen/vscode/lammps_FDM/test_fdm/1A_protein_lib/Villin_headpiece/openmm/one_start_MSM/pyemma/~/whshen/anaconda3/envs/workshop/lib/python3.9/site-packages/deeptime/markov/hmm/_bayesian_hmm.py:65), in BayesianHMMPosterior.submodel_largest(self, directed, connectivity_threshold, observe_nonempty, dtrajs)
     44 def submodel_largest(self, directed=True, connectivity_threshold='1/n', observe_nonempty=True, dtrajs=None):
     45     r""" Creates a submodel from the largest connected set.
     46 
     47     Parameters
   (...)
     63         The submodel.
     64     """
---> 65     dtrajs = ensure_dtraj_list(dtrajs)
     66     states = self.prior.states_largest(directed=directed, connectivity_threshold=connectivity_threshold)
     67     obs = self.prior.nonempty_obs(dtrajs) if observe_nonempty else None

File [~/whshen/anaconda3/envs/workshop/lib/python3.9/site-packages/deeptime/util/types.py:125](https://vscode-remote+ssh-002dremote-002b10-002e2-002e1-002e240.vscode-resource.vscode-cdn.net/home/xhshi/whshen/vscode/lammps_FDM/test_fdm/1A_protein_lib/Villin_headpiece/openmm/one_start_MSM/pyemma/~/whshen/anaconda3/envs/workshop/lib/python3.9/site-packages/deeptime/util/types.py:125), in ensure_dtraj_list(dtrajs)
    123 def ensure_dtraj_list(dtrajs):
    124     """Makes sure that dtrajs is a list of discrete trajectories (array of int)"""
--> 125     if len(dtrajs) > 0 and isinstance(dtrajs[0], numbers.Real):
    126         return [ensure_integer_array(dtrajs)]
    127     return [ensure_integer_array(t) for t in dtrajs]

TypeError: object of type 'NoneType' has no len()`

**Comment by ShenWenHuibit** [comment]
> It most likely means that due to the sampling you have produced a few hmm's who live on different states (or at least not a subset) of the prior's states. I suggest you subselect the samples accordingly. It might even be enough to call `bhmm_largest = bhmm.submodel_largest()`

When I finished running the command `bhmm_largest = bhmm.submodel_largest()` ,it still gave me the same error, and I noticed that the notes on metastable_sets in the manual read: "This is only recommended for visualization purposes. You cannot compute any actual quantity of the coarse-grained kinetics without employing the fuzzy memberships!"
I think the problem may be here, I can't directly use `bhmm_largest.prior.metastable_sets[i]` to get MFPT.

**Comment by clonker** [comment]
By the way! We gave a workshop on deeptime/pyemma a while ago and also covered HMMs there. You can find the corresponding notebook here. Given a few free minutes I'll integrate it into the documentation here, I think. :)

https://github.com/markovmodel/pyemma-workshop/blob/master/notebooks/06-hmm.ipynb

**Comment by ShenWenHuibit** [comment]
That's right, I understand now, but I have encountered another problem, please ask for your advice, it seems to be a problem with my data sample, the MFPT I calculated is too large. At the same time give me the following warning: `LinAlgWarning: Ill-conditioned matrix (rcond=1.40001e-17): result may not be accurate.` But I can get reasonable-looking results with BayesianMSM, which makes me very puzzled.

**Comment by clonker** [comment]
It could be that the transitions are not sampled well enough, but I am guessing here. Have you tried other combinations of metastable states? It could also be a problem with your clustering, projection method, featurization.... MSM and HMM estimation can be tricky. I suggest you methodically check everything and also score it with eg VAMP-2 score. And yeah, that rcond number does not inspire confidence. You could also look at your transition matrix (of the CG matrix) to see how it looks like. And visualizing the population of data frames onto a 2d projection of your data (don't forget to check all relevant projections that fall out of the projection method of your choice). Relevance is correlated to the singular (or eigen) value of the projection component.

**Comment by clonker** [comment]
And if you get reasonable results with bayesian msm you may have to tweak the estimation parameters a little when it comes to the bhmm. The prior is estimated with lower precision than the 'normal' bhmm.
